### PR TITLE
Mv drag and drop

### PR DESCRIPTION
**Pull request by huwshimi**
Added drag and drop for placing units onto machines and containers.

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- use `/:flags:/mv/`
- add a service
- open the machine view
- add some machines and containers
- check that you can drag and drop units on the machines and containers

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3959/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3960/
Test FAILed.

**Comment by urosj** [comment]
QA was :ok: 

Additional comment here: https://github.com/juju/juju-gui/issues/1109

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3961/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3962/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3963/
Test PASSed.

File: jujugui/static/gui/src/app/components/machine-view/machine-view.js [hunk 1]
```diff
@@ -452,6 +466,8 @@ YUI.add('machine-view', function() {
       );
     }
   });
+  juju.components.MachineView = ReactDnD.DragDropContext(
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine-view.js]: can you add a line above/below, give it some room to breath :)

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js [hunk 2]
```diff
@@ -20,11 +20,26 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 YUI.add('machine-view-machine', function() {
 
-  juju.components.MachineViewMachine = React.createClass({
+  var dropTarget = {
+    drop: function (props, monitor, component) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: This is kind of cryptic, could you add some doc blocs to these methods adding a link to the official docs.

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js [hunk 3]
```diff
@@ -20,11 +20,26 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+      var item = monitor.getItem();
+      props.dropUnit(item.unit, props.machine.id);
+    }
+  };
+
+  function collect(connect, monitor) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: see above about docs

File: jujugui/static/gui/src/app/components/machine-view/machine/machine.js [hunk 4]
```diff
@@ -188,10 +202,17 @@ YUI.add('machine-view-machine', function() {
...
+          <div className="machine-view__machine-drop-target">
+            <div className="machine-view__machine-drop-message">
+              Add to {this.props.machine.displayName}
+            </div>
+          </div>
         </div>
       );
     }
   });
+  juju.components.MachineViewMachine = ReactDnD.DropTarget(
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine/machine.js]: breeeeaaaath

File: jujugui/static/gui/src/app/components/machine-view/machine/test-machine.js [hunk 5]
```diff
@@ -55,7 +55,8 @@ describe('MachineViewMachine', function() {
       })
     };
     var renderer = jsTestUtils.shallowRender(
-      <juju.components.MachineViewMachine
+      <juju.components.MachineViewMachine.DecoratedComponent
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/machine/test-machine.js]: maybe a comment about this `DecoratedComponent`

File: jujugui/static/gui/src/app/components/machine-view/unplaced-unit/unplaced-unit.js [hunk 6]
```diff
@@ -19,8 +19,20 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
 'use strict';
 
 YUI.add('machine-view-unplaced-unit', function() {
+  var dragSource = {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/unplaced-unit/unplaced-unit.js]: see above about docs

File: jujugui/static/gui/src/app/components/machine-view/unplaced-unit/unplaced-unit.js [hunk 7]
```diff
@@ -76,18 +101,21 @@ YUI.add('machine-view-unplaced-unit', function() {
...
           {unit.displayName}
           <juju.components.MoreMenu
             items={menuItems} />
           {this._generatePlaceUnit()}
+          <div className="machine-view__unplaced-unit-drag-state"></div>
         </li>
       );
     }
   });
+  juju.components.MachineViewUnplacedUnit = ReactDnD.DragSource(
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/machine-view/unplaced-unit/unplaced-unit.js]: brrrrrreeaath

**Comment by hatched** [comment]
:+1: :ok: 
This is so much awesome!

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3972/
Test PASSed.

**Comment by huwshimi** [comment]
Thanks @hatched :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**